webapp/components/interop.js

```diff
@@ -258,39 +260,108 @@ class WPTInterop extends QueryBuilder(
     super.connectedCallback();
-    this.shadowRoot.querySelector('test-search')
-      .addEventListener('commit', this.onSearchCommit);
-    this.shadowRoot.querySelector('test-search')
-      .addEventListener('autocomplete', this.onSearchAutocomplete);
+    this.testSearch.addEventListener('commit', this.onSearchCommit);
+    this.testSearch.addEventListener('autocomplete', this.onSearchAutocomplete);
   }
 
   disconnectedCallback() {
-    this.shadowRoot.querySelector('test-search')
-      .removeEventListener('commit', this.onSearchCommit);
+    this.testSearch.removeEventListener('commit', this.onSearchCommit);
     super.disconnectedCallback();
   }
 
   async ready() {
     await super.ready();
+    this._createMethodObserver('precomputedInteropLoaded(precomputedInterop)');
+    if (this.structuredQueries && this.searchCacheInterop) {
+      this.fetchSearchCacheInterop();
+    } else {
+      this.fetchPrecomputedInterop();
+    }
+  }
+
+  get testSearch() {
+    return this.shadowRoot.querySelector('test-search');
+  }
+
+  fetchPrecomputedInterop() {
     this.load(
       fetch(`/api/interop${this.query}`)
         .then(async r => {
           if (!r.ok || r.status !== 200) {
             Promise.reject('Failed to fetch interop data');
           }
-          this.passRateMetadata = await r.json();
-          this.allMetrics = await fetch(this.passRateMetadata.url).then(r => r.json());
+          const metadata = await r.json();
+          this.passRateMetadata = metadata;
+          this.testRuns = metadata && metadata.test_runs;
+          this.precomputedInterop = await fetch(this.passRateMetadata.url).then(r => r.json());
+        })
+        .then(() => {
+          if (this.search) {
+            this.handleSearchCommit(this.search);
+          }
+        })
+    );
+  }
+
+  fetchSearchCacheInterop() {
+    this.load(
+      Promise.resolve(this.testRuns || this.loadRuns())
+        .then(runs => {
+          if (!runs) {
+            return;
+          }
+          const body = {
+            run_ids: this.testRuns.map(r => r.id),
+          };
+          if (this.structuredSearch) {
+            body.query = this.structuredSearch;
+          }
+          let url = new URL('/api/search', window.location);
+          url.searchParams.set('interop', ''); // Include interop scores
+          let fetchOpts = {
+            method: 'POST',
+            body: JSON.stringify(body),
+          };
+
+          // Fetch search results and refresh display nodes. If fetch error is HTTP'
+          // 422, expect backend to attempt write-on-read of missing data. In such
+          // cases, retry fetch up to 5 times with 5000ms waits in between.
+          const toast = this.shadowRoot.querySelector('#runsNotInCache');
+          return this.retry(
+            async() => {
+              const r = await window.fetch(url, fetchOpts);
+              if (!r.ok) {
+                if (fetchOpts.method === 'POST' && r.status === 422) {
```

Comment:
Same question about `fetchOpts.method` here as in https://github.com/web-platform-tests/wpt.fyi/pull/1103#discussion_r253500441. This part of the condition will always be true.

Reply:
Done.